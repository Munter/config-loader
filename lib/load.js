const chalk = require('chalk');
const mock = require('mock-require');
const resolvePath = require('resolve').sync;
const webpackLog = require('webpack-log');

const LoadConfigError = require('./LoadConfigError');
const RequireModuleError = require('./RequireModuleError');

module.exports = (argv, options = {}) => {
  const altExtensions = [null, '.es6', '.flow', '.mjs', '.ts'];
  const cosmicOptions = {
    rcExtensions: true,
    sync: true,
  };
  const configPrefix = 'webpack';
  const cwd = process.cwd();
  const log = webpackLog({ name: 'config', id: 'webpack-config-loader' });
  const requires = [].concat(argv.require).filter((r) => !!r);

  // eslint-disable-next-line no-param-reassign
  options = Object.assign({ cwd: process.cwd() }, options);

  for (const module of requires) {
    const modulePath = resolvePath(module, { basedir: cwd });

    log.info(chalk`Requiring the {cyan ${module}} module`);

    try {
      if (argv.requireOptions) {
        const requireOptions = JSON.parse(argv.requireOptions);
        // eslint-disable-next-line import/no-dynamic-require, global-require
        require(modulePath)(requireOptions);
      } else {
        // eslint-disable-next-line import/no-dynamic-require, global-require
        require(modulePath);
      }
    } catch (e) {
      log.error(chalk`An error ocurred while requiring: {grey ${module}}`);
      throw new RequireModuleError(e, module);
    }
  }

  mock('require-from-string', (...args) => {
    const [, path] = args;
    const resolved = resolvePath(path, { basedir: cwd });

    // eslint-disable-next-line import/no-dynamic-require, global-require
    return require(resolved);
  });

  // ensures that cosmiconfig is using our mock
  const cosmiconfig = mock.reRequire('cosmiconfig');
  let config;
  let configPath;

  for (const extension of altExtensions) {
    let js = `${configPrefix}.config.js`;

    if (extension) {
      js = `${configPrefix}.config${extension}`;
      log.info(chalk`Config not found, trying {grey ${js}}`);
    }

    const opts = Object.assign({}, cosmicOptions, { js });
    const explorer = cosmiconfig(configPrefix, opts);

    try {
      ({ config, filepath: configPath } = explorer.load(options.cwd) || {});
    } catch (e) {
      // prettier-ignore
      log.error(chalk`An error ocurred while trying to load: {grey ${configPath || js}}
            Did you forget to specify a --require?`);
      throw new LoadConfigError(e, configPath);
    }

    if (config) {
      break;
    }
  }

  mock.stop('require-from-string');

  if (!configPath) {
    // prettier-ignore
    log.error(chalk`Unable to load a config from: {grey ${options.cwd}}`);
    const e = new Error(`Unable to load a config from: ${options.cwd}`);
    throw new LoadConfigError(e, configPath);
  }

  return { config, configPath };
};
