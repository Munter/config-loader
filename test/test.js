require('loud-rejection/register');
require('./tests/load');
