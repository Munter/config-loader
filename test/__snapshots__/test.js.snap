// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load should config of type common-js 1`] = `
Object {
  "mode": "common-js",
}
`;

exports[`Load should config of type es6 1`] = `
Object {
  "default": Object {
    "mode": "es6",
  },
}
`;

exports[`Load should config of type flow 1`] = `
Object {
  "mode": "flow",
}
`;

exports[`Load should config of type typescript 1`] = `
Object {
  "default": Object {
    "mode": "typescript",
  },
}
`;
